include/field3d.hxx

```diff
@@ -226,30 +228,42 @@ class Field3D : public Field, public FieldData {
   
   /// Check if this field has yup and ydown fields
   bool hasYupYdown() const {
-    return (yup_field != nullptr) && (ydown_field != nullptr);
+    return !yup_fields.empty() and !ydown_fields.empty();
   }
 
   /// Return reference to yup field
-  Field3D& yup() { 
-    ASSERT2(yup_field != nullptr); // Check for communicate
-    return *yup_field; 
+  Field3D &yup(std::vector<Field3D>::size_type index = 0) {
+    if (yup_fields.empty()) {
+      return *this;
```

Comment:
Do we want to return `*this` rather than throwing an exception? Might cause confusion if `ShiftedMetric` is being used, but the user calls `yup()` before `calcYUpDown()`.

Actually, maybe this needs a `bool hasValidYUpDown` type check. If `mergeYupYdown()` has been called and `hasValidYUpDown=true` has been set, we want to return `*this`, but otherwise it would be an error.

Reply:
Yes, this was initially an optimisation for the identity transform, but just copying the field into the parallel slice vectors should be sufficient -- COW should prevent this being too costly.

I'm currently thinking we can have a `bool hasValidParallelSlices` that the non-`const` methods set to `false`, with a public `setParallelSlicesValid` (or some less horrible name) that sets it to `true`. The parallel transforms would call that, as well as possibly advanced users if they what they're doing.